doc/rtd/explanation/format.rst

```diff
@@ -24,64 +69,113 @@ These things may include:
+module to determine the run frequency.
 
-.. note::
-   Cloud config data can also render cloud instance metadata variables using
-   :ref:`jinja templates <instancedata-Using>`.
+See the :ref:`yaml_examples` section for a set of commented examples of
+supported cloud config formats.
 
 .. _user_data_script:
 
 User data script
 ================
 
-Typically used by those who just want to execute a shell script.
+| **Header:** #!
+| **Content-Type:** text/x-shellscript
 
-Begins with: ``#!`` or ``Content-Type: text/x-shellscript`` when using a MIME
-archive.
+**Example**
 
-User data scripts can optionally render cloud instance metadata variables using
-:ref:`jinja templates <instancedata-Using>`.
+.. code-block:: shell
 
-Example script
---------------
+    #!/bin/sh
+    echo "Hello World" > /var/tmp/output.txt
 
-Create a script file :file:`myscript.sh` that contains the following:
+**Explanation**
 
-.. code-block::
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, after most
+other cloud-init modules have run.
 
+.. _user_data_formats-cloud_boothook:
+
+Cloud boothook
+==============
+
+| **Header:** #cloud-boothook
+| **Content-Type:** text/cloud-boothook
+
+**Simple Example**
+
+.. code-block:: shell
+
+   #cloud-boothook
    #!/bin/sh
-   echo "Hello World.  The time is now $(date -R)!" | tee /root/output.txt
+   echo 192.168.1.130 us.archive.ubuntu.com > /etc/hosts
 
-Now run:
+**Example of once-per-instance script**
 
-.. code-block:: shell-session
+.. code-block:: bash
 
-   $ euca-run-instances --key mykey --user-data-file myscript.sh ami-a07d95c9
+   #cloud-boothook
+   #!/bin/sh
 
-Kernel command line
-===================
+   PERSIST_ID=/var/lib/cloud/first-instance-id
+   _id=""
+   if [ -r $PERSIST_ID ]; then
+     _id=$(cat /var/lib/cloud/first-instance-id)
+   fi
 
-When using the NoCloud datasource, users can pass user data via the kernel
-command line parameters. See the :ref:`NoCloud datasource<datasource_nocloud>`
-and :ref:`explanation/kernel-command-line:Kernel command line` documentation
-for more details.
+   if [ -z $_id ]  || [ $INSTANCE_ID != $_id ]; then
+     echo 192.168.1.130 us.archive.ubuntu.com >> /etc/hosts
+   fi
+   sudo echo $INSTANCE_ID > $PERSIST_ID
 
-Gzip compressed content
-=======================
+**Explanation**
 
-Content found to be gzip compressed will be uncompressed.
-The uncompressed data will then be used as if it were not compressed.
-This is typically useful because user data is limited to ~16384 [#]_ bytes.
+A cloud boothook is similar to a :ref:`user data script<user_data_script>`
+in that it is a shell script run on boot. The boothook is different in that:
+
+* It is run very early in boot, even before the ``cc_bootcmd`` module
+* It is run on every boot
+* The environment variable ``INSTANCE_ID`` is set to the current instance ID
```

Comment:
`INSTANCE_ID` is also provided for [cc_bootcmd](https://github.com/canonical/cloud-init/blob/779dd6b009f8a568fe74e98df9803ad590b83044/doc/module-docs/cc_bootcmd/data.yaml#L5), could you please remove this item?